k-distribution/src/main/scripts/lib/pyk/pyk/cterm.py

```diff
@@ -0,0 +1,105 @@
+)
+from .prelude import mlAnd
+from .subst import Subst
+
+
+@dataclass(frozen=True)
+class CTerm:
+    config: KInner  # TODO Optional?
+    constraints: Tuple[KInner, ...]
+
+    def __init__(self, cterm: KInner) -> None:
+        config, constraint = splitConfigAndConstraints(cterm)
+        constraints = tuple(flattenLabel('#And', constraint))
+        object.__setattr__(self, 'config', config)
+        object.__setattr__(self, 'constraints', constraints)
+
+    def __iter__(self):
+        return chain([self.config], self.constraints)
+
+    @cached_property
+    def cterm(self) -> KInner:
+        return mlAnd(self)
+
+    @property
+    def hash(self) -> str:
+        return self.cterm.hash
+
+    @staticmethod
+    def from_rule(rule: KRuleLike) -> Tuple['CTerm', 'CTerm']:
+        body = drop_var_prefixes(rule.body)
+        lhs = extract_lhs(body)
+        rhs = extract_rhs(body)
+        requires = drop_var_prefixes(rule.requires)
+        ensures = drop_var_prefixes(rule.ensures)
+        return CTerm(mlAnd([lhs, requires])), CTerm(mlAnd([rhs, ensures]))
+
+    def to_rule(self, other: 'CTerm', path_constraints: Iterable[KInner] = [], *, claim=False) -> KRuleLike:
+        lhs_constraints = set(chain(self.constraints, path_constraints))
+        rhs_constraints = set(c for c in other.constraints if c not in lhs_constraints)
+
+        body = pushDownRewrites(KRewrite(self.config, other.config))
+        requires = simplifyBool(unsafeMlPredToBool(mlAnd(lhs_constraints)))
+        ensures = simplifyBool(unsafeMlPredToBool(mlAnd(rhs_constraints)))
+
+        var_count = count_vars(body) + count_vars(requires) + count_vars(ensures)
+        rhs_var_count = count_rhs_vars(body) + count_vars(ensures)
+
+        singletons = {k for k, v in var_count.items() if v == 1}
+        existentials = {k for k, v in rhs_var_count.items() if var_count[k] <= v}
+
+        subst_items = ((k, '_' + k) for k in singletons)
+        subst_items = ((k, ('?' if k in existentials else '') + v) for k, v in subst_items)
+        subst_dict = {k: KVariable(v) for k, v in subst_items}
+        subst = Subst(subst_dict)
+
+        body = subst(body)
+        requires = subst(requires)
+        ensures = subst(ensures)
+
+        if claim:
+            return KClaim(body=body, requires=requires, ensures=ensures)
+
+        return KRule(body=body, requires=requires, ensures=ensures)
+
+    def match(self, pattern: 'CTerm') -> Optional[Tuple[Subst, Set[KInner], Set[KInner]]]:
+        subst = match(pattern=pattern.config, term=self.config)
+
+        if subst is None:
+            return None
+
+        subst = Subst(subst)
+
+        antecedents = set(flattenLabel('_andBool_', simplifyBool(unsafeMlPredToBool(mlAnd(self.constraints)))))
+        consequents = {
+            conjunct
+            for conjunct in flattenLabel('_andBool_', simplifyBool(unsafeMlPredToBool(subst(mlAnd(pattern.constraints)))))
+            if conjunct not in antecedents
+        }
```

Comment:
We still should not be bringing the `Bool` theory into this module, it's too specific for these use cases.

A `CTerm` is a _Matching Logic Pattern_ with a specific form, not a _term_ with a set of _boolean constraints_.

So converting to bool just to apply simplifications is a bad idea. We should not have `Bool` leak into this module at all.

I think you can accomplish the same thing without converting it to `Bool` sort.

Reply:
If later we find it's necessary for some reason, then we can do it. But we really should be _very_ careful about using `unsafeMlPredToBool` where it's not strictly needed, it is actually unsafe.

Reply:
Pushed a change, please check.

> We should not have Bool leak into this module at all.

Does this apply to `CTerm.to_rule` as well?

Reply:
`CTerm.to_rule` and `CTerm.from_rule` removed for now.